File: src/test/java/de/fraunhofer/aisec/crymlin/CommandsTest.java

package de.fraunhofer.aisec.crymlin;

import de.fraunhofer.aisec.analysis.Commands;
import de.fraunhofer.aisec.analysis.JythonInterpreter;
import de.fraunhofer.aisec.analysis.server.AnalysisServer;
import de.fraunhofer.aisec.analysis.structures.ServerConfiguration;
import de.fraunhofer.aisec.analysis.structures.TypestateMode;
import org.junit.jupiter.api.Test;

import java.io.*;

import static org.junit.jupiter.api.Assertions.*;

/** Testing console commands. */
class CommandsTest {

	@Test
	void commandConsoleTest() {
		AnalysisServer server = AnalysisServer.builder()
				.config(ServerConfiguration.builder()
						.launchLsp(false)
						.launchConsole(true)
						.typestateAnalysis(TypestateMode.NFA)
						.markFiles(new File("src/test/resources/mark_java").getAbsolutePath())
						.useLegacyEvaluator(true)
						.build())
				.build();

		server.start();
		PrintStream oldOut = System.out;
		PrintStream oldErr = System.err;
		ByteArrayOutputStream bosOut = new ByteArrayOutputStream();
		PrintStream o = new PrintStream(bosOut);
		ByteArrayOutputStream bosErr = new ByteArrayOutputStream();
		PrintStream e = new PrintStream(bosErr);

		System.setOut(o);
		System.setErr(e);

		// Init console
		JythonInterpreter interp = new JythonInterpreter();
		Commands com = new Commands(interp);

		Commands.help();
		assertTrue(bosOut.toString().contains("help"));

		com.load_rules("src/test/resources/mark_java");

		com.list_rules();
		assertTrue(bosOut.toString().contains("Foo"));

		com.analyze("src/test/resources/good/Bouncycastle.java");
		com.show_findings();

		System.setOut(oldOut);
		System.setErr(oldErr);
	}
}

File: src/test/java/de/fraunhofer/aisec/codyze/legacy/crymlin/CommandsTest.kt
package de.fraunhofer.aisec.codyze.legacy.crymlin

import de.fraunhofer.aisec.codyze.legacy.Commands
import de.fraunhofer.aisec.codyze.legacy.JythonInterpreter
import de.fraunhofer.aisec.codyze.legacy.analysis.AnalysisServer
import de.fraunhofer.aisec.codyze.legacy.analysis.TypestateMode
import de.fraunhofer.aisec.codyze.legacy.config.CodyzeConfiguration
import de.fraunhofer.aisec.codyze.legacy.config.Configuration
import de.fraunhofer.aisec.codyze.legacy.config.CpgConfiguration
import java.io.*
import kotlin.test.assertTrue
import org.junit.jupiter.api.*

/** Testing console commands. */
internal class CommandsTest {
    @Test
    fun commandConsoleTest() {
        val codyze = CodyzeConfiguration()
        codyze.analysis.tsMode = TypestateMode.DFA
        codyze.mark = arrayOf(File("src/test/resources/legacy/mark_java"))

        val config = Configuration(codyze, CpgConfiguration())
        config.executionMode.isCli = true
        config.executionMode.isLsp = false

        val server = AnalysisServer(config)
        server.start()
        val oldOut = System.out
        val oldErr = System.err
        val bosOut = ByteArrayOutputStream()
        val o = PrintStream(bosOut)
        val bosErr = ByteArrayOutputStream()
        val e = PrintStream(bosErr)
        System.setOut(o)
        System.setErr(e)

        // Init console
        val interp = JythonInterpreter()
        val com = Commands(interp)
        Commands.help()
        assertTrue(bosOut.toString().contains("help"))

        com.load_rules("src/test/resources/legacy/mark_java")
        com.list_rules()
        assertTrue(bosOut.toString().contains("Foo"))

        com.analyze("src/test/resources/legacy/good/Bouncycastle.java")
        com.show_findings()
        System.setOut(oldOut)
        System.setErr(oldErr)
    }
}
